src/schemapack/_internals/spec/custom_types.py

```diff
@@ -26,3 +27,5 @@
 RelationPropertyName: TypeAlias = _NonEmptyStr
 ContentPropertyName: TypeAlias = _NonEmptyStr
 IdPropertyName: TypeAlias = _NonEmptyStr
+FrozenType: TypeAlias = FrozenDict[str, str | FrozenDict | tuple]
+ThawedType: TypeAlias = dict[str, str | dict | list]
```

Comment:
There's a way to actually have recursive type definitions here using forward references:
```suggestion
FrozenType: TypeAlias = FrozenDict[str, "str | FrozenType | tuple"]
ThawedType: TypeAlias = dict[str, "str | ThawedType | list"]
```
See https://docs.python.org/3/library/stdtypes.html#union-type for more details on why the whole union has to be wrapped in a string in this case.